### PR TITLE
VDS-323: Put customer segment user group into index

### PR DESCRIPTION
**Pull request by asvishnyakov**
Depends on https://github.com/VirtoCommerce/vc-module-demo-customer-segments/pull/12

### CODE REVIEW

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: tests/VirtoCommerce.DemoCustomerSegmentsModule.Tests/CustomerSegmentExpressionTest.cs [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+            var firstCustomerUserGroups = GetUserGroups(documents[0]);
+            Assert.Equal("Test", firstCustomerUserGroups.ToString());
+
+            var secondCustomerUserGroups = GetUserGroups(documents[1]);
+            Assert.Null(secondCustomerUserGroups);
+        }
+
+        [Fact]
+        public async Task Multiple_Stores()
```
Comment (pushnitsa) [inline tests/VirtoCommerce.DemoCustomerSegmentsModule.Tests/CustomerSegmentExpressionTest.cs:137]: Please fix tests formatting. 
Group it by
// Arrange
// Act
// Assert
Like other VC tests

Comment (pushnitsa) [reply tests/VirtoCommerce.DemoCustomerSegmentsModule.Tests/CustomerSegmentExpressionTest.cs:137]: Also tests should be named according microsoft recommendations
https://docs.microsoft.com/ru-ru/dotnet/core/testing/unit-testing-best-practices#naming-your-tests

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+            if (member is IHasSecurityAccounts hasSecurityAccounts)
+            {
+                document.AddFilterableAndSearchableValues("Stores", hasSecurityAccounts.SecurityAccounts.Select(x => x.StoreId).ToArray());
+            }
+
+            if (member is Contact customer)
+            {
+                var evaluationContext = new DemoCustomerSegmentExpressionEvaluationContext { Customer = customer };
+                var userGroups = _customerSegments
```
[the remark this answers is not in the document]
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentBuilder.cs]: Fixed with workaround

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentChangesProvider.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+        public override async Task<long> GetTotalChangesCountAsync(DateTime? startDate, DateTime? endDate)
+        {
+            long result;
+
+            var customerSegmentsCount = await GetCustomerSegmentsChangesCount(startDate, endDate);
+            // If any customer segments changed, then all members should be indexed
+            if (customerSegmentsCount > 0)
+            {
+                // Get total members count
```
[the remark this answers is not in the document]
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentChangesProvider.cs]: Removed

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentChangesProvider.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            _memberRepositoryFactory = memberRepositoryFactory;
+            _changeLogSearchService = changeLogSearchService;
+        }
+
+        public override async Task<long> GetTotalChangesCountAsync(DateTime? startDate, DateTime? endDate)
+        {
+            long result;
+
+            var customerSegmentsCount = await GetCustomerSegmentsChangesCount(startDate, endDate);
+            // If any customer segments changed, then all members should be indexed
```
[the remark this answers is not in the document]
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentChangesProvider.cs]: Removed, renamed

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Module.cs [hunk 5]
```diff
@@ -42,6 +56,29 @@ public void PostInitialize(IApplicationBuilder appBuilder)
             AbstractTypeFactory<IConditionTree>.RegisterType<DemoBlockCustomerSegmentRule>();
             AbstractTypeFactory<IConditionTree>.RegisterType<DemoConditionPropertyValues>();
 
+            var inProcessBus = appBuilder.ApplicationServices.GetService<IHandlerRegistrar>();
+            inProcessBus.RegisterHandler<DemoCustomerSegmentChangedEvent>(async (message, token) => await appBuilder.ApplicationServices.GetService<LogChangesEventHandler>().Handle(message));
+            inProcessBus.RegisterHandler<DemoCustomerSegmentChangedEvent>(async (message, token) => await appBuilder.ApplicationServices.GetService<CacheCustomerSegmentChangedEventHandler>().Handle(message));
+
+            var settingsManager = appBuilder.ApplicationServices.GetService<ISettingsManager>();
+            if (settingsManager.GetValue(ModuleConstants.Settings.General.EventBasedIndexation.Name, false))
```
[the remark this answers is not in the document]
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoCustomerSegmentsModule.Web/Module.cs]: Done

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentChangesProvider.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            if (customerSegmentsCount > 0)
+            {
+                // Get total members count
+                using var memberRepository = _memberRepositoryFactory();
+                result = memberRepository.Members.Count();
```
[the remark this answers is not in the document]
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentChangesProvider.cs]: Fixed here in in `ToArray` below

File: src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Services/UserGroupEvaluator.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    public class UserGroupEvaluator: IUserGroupEvaluator
+    {
+        private readonly IDemoCustomerSegmentSearchService _customerSegmentSearchService;
+
+        public UserGroupEvaluator(IDemoCustomerSegmentSearchService customerSegmentSearchService)
+        {
+            _customerSegmentSearchService = customerSegmentSearchService;
+        }
+
+        public async Task<ICollection<string>> EvaluateUserGroupsAsync(IEvaluationContext context)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Services/UserGroupEvaluator.cs:21]: I think we also need a same synchronous method here.

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: tests/VirtoCommerce.DemoCustomerSegmentsModule.Tests/CustomerSegmentExpressionTest.cs (hunk 1 above)
Comment (asvishnyakov) [reply tests/VirtoCommerce.DemoCustomerSegmentsModule.Tests/CustomerSegmentExpressionTest.cs]: Refactored

**Review by pushnitsa — APPROVED** [review]

**Pull request merged**